lib/pbbuilder/template.rb

```diff
@@ -0,0 +1,91 @@
+# PbbuilderTemplate is an extension of Pbbuilder to be used as a Rails template
+# It adds support for partials.
+class PbbuilderTemplate < Pbbuilder
+  class << self
+    attr_accessor :template_lookup_options
+  end
+
+  self.template_lookup_options = {handlers: [:pbbuilder]}
+
+  def initialize(context, message)
+    @context = context
+    super(message)
+  end
+
+  # Render a partial. Can be called as:
+  # pb.partial! "name/of_partial", argument: 123
+  # pb.partial! "name/of_partial", locals: {argument: 123}
+  # pb.partial! partial: "name/of_partial", argument: 123
+  # pb.partial! partial: "name/of_partial", locals: {argument: 123}
+  # pb.partial! @model # @model is an ActiveModel value, it will use the name to look up a partial
+  # pb.friends @friends, partial: "friend", as: :friend # Render a collection by using a partial for each element
+  def partial!(*args)
+    if args.one? && _is_active_model?(args.first)
+      _render_active_model_partial args.first
+    else
+      _render_explicit_partial(*args)
+    end
+  end
+
+  def set!(field, *args, **kwargs, &block)
+    if args.one? && _partial_options?(kwargs)
+      super(field, *args) do |element|
+        _set_inline_partial(element, kwargs)
+      end
+    else
+      super
+    end
+  end
+
+  private
+
+  def _partial_options?(options)
+    ::Hash === options && options.key?(:as) && options.key?(:partial)
+  end
+
+  def _is_active_model?(object)
+    object.class.respond_to?(:model_name) && object.respond_to?(:to_partial_path)
```

Comment:
I think it may be possible for an object to inherit `ActiveModel::Name`, but not `ActiveModel::Conversion`. What's the thinking around not only checking if it responds to `to_partial_path`?

Reply:
This is copied from jbuilder https://github.com/rails/jbuilder/blob/770abdb6b4f027ef577c16fe1f88bf0ed2db46e4/lib/jbuilder/jbuilder_template.rb#L180-L182